### DISCUSSION TITLE
Check code as if in `PYTHONOPTIMIZE` mode

### DISCUSSION THREAD

**Discussion by Zac-HD** [pschanely/CrossHair#120] [Ideas]
If [run with `python -O ...`](https://docs.python.org/3/using/cmdline.html#cmdoption-O), Python will strip out any `assert` statements and code dependent on the `__debug__` constant at compile time.  This occasionally causes bugs when users wrongly assume that such code will always be executed; some [linters even recommend avoiding `assert`](https://bandit.readthedocs.io/en/latest/plugins/b101_assert_used.html) which is IMO an overreaction.

On one hand it's also pretty easy to check for: just run your code with `-O` and `-OO` in addition to normal mode; on the other almost nobody actually does this... and on the gripping hand maybe it would be easy for `crosshair` to add an implicit branch on this?  It's at most a 3x penalty, and I'd expect usually a *lot* less as most code doesn't depend on that value.

Taking the idea further, people also claim that for this reason `assert` statements should be side-effect-free, which IMO is too strict a constraint: we only need them to be *morally pure*, i.e. not have invariant-violating side effects (whereas e.g. populating a cache is fine).  I think a tool to disable a random subset of assertions should probably be a separate thing though, and can then be integrated into crosshair and/or a pytest plugin.

**Comment by pschanely** [comment]
Zac: I thought about a response to this for quite a while and then completely forgot about it. Apologies.

So, first: this idea is clever and seems at least moderately valuable.

Sadly, branching on `-O` behavior is actually sort of challenging for CrossHair: as you said, the "optimization" happens [during bytecode generation](https://stackoverflow.com/a/28608386/3654346), and CrossHair runs in a fairly dynamic manner today. `crosshair watch` may have a slight advantage here, as it runs a pool of subprocesses: perhaps we could just run some of them with optimization turned on.

Just to confirm, for the first version, were you imagining that we'd check that we don't violate any contracts (or as usual, raise unexpected exceptions) under `-O`?

A much more minor stumbling block is that we don't have a pattern for describing the counterexample. CrossHair has evolved some patterns for counterexamples-under-context-managers, e.g. `patch_to_return()` in this [how-to](https://crosshair.readthedocs.io/en/latest/case_studies.html). Python is so dynamic that that `-O` might be one of the only cases where that pattern won't work.

**Reply by Zac-HD** [reply]
> Just to confirm, for the first version, were you imagining that we'd check that we don't violate any contracts (or as usual, raise unexpected exceptions) under `-O`?

Yes, that (and `-OO`) is the whole suggestion!  I don't have much of a suggestion for how to display this though 😅

**Reply by pschanely** [reply]
> > Just to confirm, for the first version, were you imagining that we'd check that we don't violate any contracts (or as usual, raise unexpected exceptions) under `-O`?
> 
> Yes, that (and `-OO`) is the whole suggestion! I don't have much of a suggestion for how to display this though 😅

Fantastic, thanks! Amusingly, I have a suspicion that `-OO` breaks crosshair itself by way of PEP316 contract parsing.  And yeah, the display side is of course solvable; it's just text after all.

**Folks on the internet**: if a feature like this existed, would you use it? Comment below. An extra vote or two would go a long way towards helping me prioritize it.
